### ISSUE TITLE
Features.

### ISSUE THREAD

**Issue by octalmage** [octalmage/ZenApp#1]
## Lets brainstorm!

To start I'm just going to implement native desktop notifications, then we can add other features as we go. 

**Ideas**
- Native desktop notifications.
- Quick reply and ticket preview.
- Full interface for editing and replying to tickets. 
  - Is this really necessary? We'll have to recreate the UI from scratch and I'm not sure what we will gain. Maybe we can just open the web UI for a ticket in a window? 

Please share your ideas! Especially features that aren't possible in the web app. 

Thanks!

**Comment by gdespirito** [comment]
Actually, i think that it's a good idea to recreate the UI in a brand new desktop app. We can just make a simple interface to view the history and post a new public/private reply.

**Comment by octalmage** [comment]
@gonzunigad good call! Viewing the ticket history is usually necessary for sending a good reply, so we should at least have this. A minimal interface shouldn't be difficult, it just won't have fancy features like apps and macros.

**Comment by Danw33** [comment]
@octalmage Why didn't I think to do this?! Great idea! :bulb: 

@gonzunigad I'd agree with that, a fresh start would be best :+1: 

I don't have much to suggest just yet, but I'll have a good think! 
Here's what I have so far, it may be pointless ideas but ideas none the less:
- Minimise to the status menu (Hide dock icon) as a configurable option
- "Offline Mode" - This might be more of a future feature; Cache certain API responses locally so that if the user looses connection they can be told they are offline, but continue to read previously viewed tickets that have been loaded on their machine previously in a read-only mode. Might be Ideal for people using this over an (unreliable) mobile data connection or on the go?

Only other thing I can think of is License - I'm assuming (seeing as though we're on a public repo) the project will be going down the open-source route: We need to decide [what license would be best](http://choosealicense.com/) and need to know if there are any rules imposed by ZenDesk for third-party applications? (and once chosen get it in [LICENSE.md](../blob/master/LICENSE.md) for the world to see ;) ) - Personally, I'd go for MIT or GPL v3 :wink: 

Good luck with the project! :+1:

**Comment by octalmage** [comment]
@Danw33 thanks for the suggestions! I was going to have the application live in the menubar with an icon. This way it's always available! This means that we're going to need an icon, I'll create a separate issue for this. 

An offline mode is a great idea! You can work on a response without an internet connection, and save the response until you reconnect.  

I usually go with GPL for end-user apps, and MIT for libraries. So GPL it is! 

Thanks again!

**Comment by jpvasquez** [comment]
Notifications are great, but I can currently work around the lack thereof using Slack. 

My #1 feature for a Desktop (Mac) app is speed. When using the website, every time I submit a response, I have to wait a couple of seconds for the next ticket to load. The same goes if I want to see the original email message or answer with a template. It's just a couple of seconds, but when you have to answer 20 support requests everyday on top of the other work involved in running an indie app, it adds up to an enervating situation.

My ideal app would:

1 - Cache all data related to open tickets on startup, including previously closed tickets from the same requesters.
2 - Allow me to work on this data set with no network delay whatsoever.
3 - Submit ticket edits in the background
4 - When I open a ticket for edition, display its data instantaneously and refresh its status on the background, offering a refresh banner if any change is detected (unlikely, but possible when multiple agents exist).

Please tell me if I can help in any way.

Thanks for the initiative!

**Comment by octalmage** [comment]
@jpvasquez thanks for the suggestions! It's clear that a fast ticket editing interface is desired, so this will probably be included in the app.

**Comment by johlym** [comment]
@octalmage If I could relegate ZD to my menu bar and free up my browser, I'd love that. 
- If I could get notifications when new tickets arrive
- when tickets are assigned to me
- when customers reply
- when agents reply
- basically all ticket updates, pretty much like what we have with our ZD <=> Slack integration. :)

**Comment by daniel020403** [comment]
Suggestions and ideas above are all good candidates for features for a basic desktop app. We just need to make sure that the app is a cross-platform one.

Would be great to get this project started ASAP. I really would love to contribute to this project.

**Comment by my2ter** [comment]
Please don't make a Java app...

**Comment by gdespirito** [comment]
Electron.io!

**Comment by octalmage** [comment]
It will for sure be written using Electron. I just don't have a lot of free time at the moment unfortunately.

**Comment by cedricve** [comment]
Electron +1

**Comment by holywen** [comment]
Do they provide any open API now? I think this might be the first thing that we need.
